hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/HandlerBalancer.java

```diff
@@ -0,0 +1,179 @@
+ * after a successful (Hazelcast) binding process.
+ *
+ *
+ */
+public class HandlerBalancer implements Runnable {
+    private final ILogger log;
+
+    private final int migrationIntervalSeconds;
+    private final MigrationStrategy strategy;
+
+    private final IOSelectorLoadCalculator inCalculator;
+    private final IOSelectorLoadCalculator outCalculator;
+
+    private volatile boolean isActive;
+
+    public HandlerBalancer(InSelectorImpl[] inSelectors, OutSelectorImpl[] outSelectors, int migrationIntervalSeconds,
+                           LoggingService loggingService) {
+        this.log = loggingService.getLogger(HandlerBalancer.class);
+        this.migrationIntervalSeconds = migrationIntervalSeconds;
+        this.strategy = new MigrationStrategy();
+
+        this.inCalculator = new IOSelectorLoadCalculator(inSelectors, loggingService);
+        this.outCalculator = new IOSelectorLoadCalculator(outSelectors, loggingService);
+
+        this.isActive = shouldStart(inSelectors, outSelectors);
+    }
+
+    public void connectionAdded(Connection connection) {
+        if (!(connection instanceof TcpIpConnection)) {
+            return;
+        }
+
+        ReadHandler readHandler = ((TcpIpConnection) connection).getReadHandler();
+        if (log.isFinestEnabled()) {
+            log.finest("Adding a read handler " + readHandler);
+        }
+        inCalculator.addHandler(readHandler);
+
+        WriteHandler writeHandler = ((TcpIpConnection) connection).getWriteHandler();
+        if (log.isFinestEnabled()) {
+            log.info("Adding a write handler " + writeHandler);
+        }
+        outCalculator.addHandler(writeHandler);
+    }
+
+    public void connectionRemoved(Connection connection) {
+        if (!(connection instanceof TcpIpConnection)) {
+            return;
+        }
+
+        ReadHandler readHandler = ((TcpIpConnection) connection).getReadHandler();
+        if (log.isFinestEnabled()) {
+            log.finest("Removing a read handler " + readHandler);
+        }
+        inCalculator.removeHandler(readHandler);
+
+        WriteHandler writeHandler = ((TcpIpConnection) connection).getWriteHandler();
+        if (log.isFinestEnabled()) {
+            log.finest("Removing a write handler " + readHandler);
+        }
+        outCalculator.removeHandler(writeHandler);
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (isActive) {
+                log.finest("Starting selector thread handler balance check");
+                checkReadHandlers();
+                checkWriteHandlers();
+                TimeUnit.SECONDS.sleep(migrationIntervalSeconds);
+            }
+        } catch (InterruptedException e) {
+            //this thread is about to exit, no reason restoring the interrupt flag
+            EmptyStatement.ignore(e);
+        }
+    }
+
+    public void stop() {
+        isActive = false;
```

Comment:
Perhaps an idea to do an interrupt as well to make sure the thread breaks free from its sleep. Check the SlowOperationDetector for an example.